storage/storage.go

```diff
@@ -185,13 +185,16 @@ func NewTupleKeyIteratorFromTupleIterator(iter TupleIterator) TupleKeyIterator {
 }
 
 // NewTupleKeyObjectIterator returns an ObjectIterator that iterates over the objects
-// contained in the provided list of TupleKeys.
-func NewTupleKeyObjectIterator(tupleKeys []*openfgapb.TupleKey) ObjectIterator {
+// contained in the provided list of TupleKeys. You can optionally pass in a function that
+// will remove the objects that don't match the filter.
+func NewTupleKeyObjectIterator(tupleKeys []*openfgapb.TupleKey, filterFunc TupleKeyFilterFunc) ObjectIterator {
```

Comment:
Please don't modify this signature. We have the [`FilteredTupleKeyIterator`](https://github.com/openfga/openfga/blob/main/storage/storage.go#L261) that serves this same purpose.

There's an example of how to do it here https://github.com/openfga/openfga/blob/main/storage/storage_test.go#L194

Reply:
The problem is that `FilteredTupleKeyIterator` returns a `TupleKeyIterator`, but in line 256 I need a `ObjectIterator`. 

I could add:

```go
func NewObjectIteratorFromTupleKeyIterator(iter TupleKeyIterator) ObjectIterator {
	objects := make([]*openfgapb.Object, 0)
	for {
		tk, err := iter.Next()
		if err == ErrIteratorDone {
			break
		}
		objectType, objectID := tuple.SplitObject(tk.GetObject())
		objects = append(objects, &openfgapb.Object{Type: objectType, Id: objectID})
	}

	return NewStaticObjectIterator(objects)
}
```

and then in ListObjects
```go
iter1 := storage.NewObjectIteratorFromTupleKeyIterator(storage.NewFilteredTupleKeyIterator(
		storage.NewStaticTupleKeyIterator(req.GetContextualTuples().GetTupleKeys()),
		func(tk *openfgapb.TupleKey) bool {
			return tuple.GetType(tk.GetObject()) == req.GetType()
		}))
iter2 := q.Datastore.ListObjectsByType(ctx, req.GetStoreId(), req.GetType())

iter := storage.NewUniqueObjectIterator(iter1, iter2)
```

Do you prefer it this way? I don't have a strong opinion.

Reply:
Yeah, I like that chaining approach better as it is more consistent with what we've been doing with the other iterator chaining.

Reply:
Updated!